[feat] 소셜 로그인 기능과 간단한 유저 서비스 구현

cjm0611 opened the pull request:
# 작업 내용
카카오 소셜 로그인
유저 정보 조회
유저 닉네임 변경
config 깃 서브모듈로 관리

# 구현 결과
- 카카오 소셜 로그인 (앱 아이콘은 임시)
<img width="601" alt="image" src="https://github.com/user-attachments/assets/c60bcd03-c1b8-4740-a4d4-70b1eebefdcf">
<img width="545" alt="image" src="https://github.com/user-attachments/assets/9c826d3e-6fb7-415f-8352-d30f18cfc964">
<img width="311" alt="image" src="https://github.com/user-attachments/assets/ba25581e-9048-4cfa-b32d-06b2c5463790">
<img width="857" alt="image" src="https://github.com/user-attachments/assets/1990b34f-a95d-416d-93fe-12e674b465c4">


http://localhost:8080/oauth2/authorization/kakao 링크로 접속하여 회원가입

<br />

- 유저 정보 조회
<img width="1075" alt="image" src="https://github.com/user-attachments/assets/7222059c-ecf9-4fc4-8ca5-637507a43382">

<br />

- 유저 닉네임 변경
<img width="1075" alt="image" src="https://github.com/user-attachments/assets/303fe704-d849-427f-b310-70b73a73611b">

# 논의 사항
1. 카카오 소셜 로그인만 제공한다면, 프로필 이미지를 필수로 받아서 연동할 수 있을 것 같은데 그렇게 진행할까요? 아니면 필수입력 항목에서 프로필 이미지를 제거하고 s3에 기본 이미지 업로드해서 해당 주소로 넣는 방법이 있을 것 같습니다.
2. User Controller Test에서 `@AuthenticationPrincipal Long id` 인자를 받아야하는 컨트롤러 메서드를 호출할 때 `given(userService.getUserInfo(any())).willReturn(expectedResponse);` given 값에 `any()`가 아닌 `anyLong()`으로 작성 시 에러가 납니다. WithMockUser 어노테이션으로는 id를 제공할 수 없어 커스텀 로직을 별도로 작성해야 하는 것 같은데, 공수가 꽤 들 것 같습니다. 우선 any()로 진행할지 아니면 커스텀 로직을 작성해야 할지 의견 궁금합니다.

cjm0611 committed:
chore: gitignore에 application.yml 추가

cjm0611 committed:
chore: OAuth2 의존성 추가

cjm0611 committed:
feat: CORS configuration 추가

cjm0611 committed:
feat: User entity 추가

cjm0611 committed:
feat: RefreshToken entity 추가

cjm0611 committed:
feat: 토큰과 쿠키 서비스 구현

cjm0611 committed:
feat: OAuth2 성공 및 실패 핸들러 구현

cjm0611 committed:
feat: 카카오 OAuth2 소셜 로그인 기능 구현

cjm0611 committed:
Merge branch 'develop' into FEAT/KAN-8

cjm0611 committed:
chore: gitignore에서 application.yml 제거

cjm0611 committed:
chore: spring-security-test 추가

cjm0611 committed:
fix: PaperControllerTest에서 불필요한 Jwt 빈 제거

cjm0611 committed:
Merge branch 'develop' into FEAT/KAN-8

cjm0611 committed:
fix: PaperLikeControllerTest Jwt 관련 오류 해결

cjm0611 committed:
fix: h2-console 요청은 토큰 검사를 수행하지 않도록 수정

cjm0611 committed:
feat: 리프레시 토큰 pk와 속성 변경

cjm0611 committed:
feat: 설정 파일 깃 서브모듈로 관리

cjm0611 committed:
chore: 설정 파일 gitignore에 추가

cjm0611 committed:
chore: 설정 파일을 추적하지 않도록 처리

cjm0611 committed:
feat: 사용자 정보, 닉네임 업데이트 기능 구현

cjm0611 committed:
feat: user 생성자에서 기본 프로필 이미지 지정

cjm0611 committed:
test: UserControllerTest 작성

cjm0611 committed:
rename: DTO suffix 변경

cjm0611 committed:
chore: csrf 비활성화

SSUHYUNKIM commented:
1. 카카오 소셜 로그인만 제공한다면, 프로필 이미지를 필수로 받아서 연동할 수 있을 것 같은데 그렇게 진행할까요? 아니면 필수입력 항목에서 프로필 이미지를 제거하고 s3에 기본 이미지 업로드해서 해당 주소로 넣는 방법이 있을 것 같습니다.
-> 프로필 이미지를 필수로 받는 과정이 사용자에게 크게 번거롭지 않을 거 같아서 필수로 받아서 연동하는 것도 좋을 거 같습니다.

2. User Controller Test에서 @AuthenticationPrincipal Long id 인자를 받아야하는 컨트롤러 메서드를 호출할 때 given(userService.getUserInfo(any())).willReturn(expectedResponse); given 값에 any()가 아닌 anyLong()으로 작성 시 에러가 납니다. WithMockUser 어노테이션으로는 id를 제공할 수 없어 커스텀 로직을 별도로 작성해야 하는 것 같은데, 공수가 꽤 들 것 같습니다. 우선 any()로 진행할지 아니면 커스텀 로직을 작성해야 할지 의견 궁금합니다.

-> 크게 문제없이 동작한다면 일단은 any()로 진행하고 추후에 방법을 같이 찾아보는게 좋을 거 같아요.
해당 이슈는 오늘 할 일에 따로 적어주세요!

SSUHYUNKIM commented on src/docs/asciidoc/api-documnet.adoc:
```diff
@@ -78,4 +78,26 @@ include::{snippets}/좋아요 상태 조회/http-request.adoc[]
+include::{snippets}/사용자 정보 조회/http-response.adoc[]
+include::{snippets}/사용자 정보 조회/response-fields.adoc[]
+
+=== 수정
```
닉네임 수정이라고 확실하게 명시하는게 나중에 빠르게 찾기 좋을 거 같아요!

SSUHYUNKIM commented on src/main/java/com/ktb/paperplebe/auth/config/CorsConfig.java:
```diff
@@ -0,0 +1,48 @@
+
+@Configuration
+public class CorsConfig {
+    @Bean(name = "corsConfigurationSource")
```
Bean을 중복해서 정의한 이유가 있을까요?

cjm0611 replied:
corsConfigurationSource 이름의 Bean을 중복 정의한 이유를 물어보시는 거라면, 개발 환경과 prod 환경에서의 config를 분리하기 위해서 정의했습니다.

SSUHYUNKIM replied:
아하 그러면 구분하기 좋게 Bean의 이름을 다르게 정의하는건 어떤가요?

SSUHYUNKIM commented on src/main/java/com/ktb/paperplebe/auth/config/CorsConfig.java:
```diff
@@ -0,0 +1,48 @@
+        CorsConfiguration configuration = new CorsConfiguration();
+
+        configuration.setAllowCredentials(true);
+        configuration.setAllowedOrigins(List.of("http://127.0.0.1:3000", "http://localhost:3000", "http://localhost:8080", "https://localhost:3000"));
```
도메인 주소는 yaml에서 관리하는 건 어떨까요?

SSUHYUNKIM commented on src/main/java/com/ktb/paperplebe/auth/config/CorsConfig.java:
```diff
@@ -0,0 +1,48 @@
+        configuration.setAllowCredentials(true);
+        configuration.setAllowedOrigins(List.of("http://127.0.0.1:3000", "http://localhost:3000", "http://localhost:8080", "https://localhost:3000"));
+        configuration.setAllowedMethods(List.of("GET", "POST", "PUT", "DELETE", "PATCH", "OPTIONS"));
+        configuration.setAllowedHeaders(List.of("*"));
```
명시적인 값을 사용하지 않고 '*'을 사용하는 이유가 있을까요?

SSUHYUNKIM commented on src/main/java/com/ktb/paperplebe/auth/config/jwt/JwtUtil.java:
```diff
@@ -0,0 +1,125 @@
+            return true;
+        } catch (ExpiredJwtException e) {
+            log.info("만료된 JWT token 입니다.");
+            return true;
```
제가 짰던 코드를 가져오신거 같은데 생각해보니까 만료가 됐는데 true를 주는게 맞을까요..?

cjm0611 replied:
false로 반환하는 게 맞겠네요!
access token이 만료된 경우 refresh token을 이용해 검사하도록 로직 수정했습니다.

cjm0611 replied:
이후에 토큰을 갱신할 때 쿠키와 리프레시 토큰을 삭제하려고 여기서 true 조건을 줬더라구요.
처음부터 access token, refresh token이 유효한지 확인하여 3가지 케이스로 분기하도록 변경했습니다!

SSUHYUNKIM commented on src/main/java/com/ktb/paperplebe/auth/dto/JwtClaimResponse.java:
```diff
@@ -0,0 +1,14 @@
+
+@AllArgsConstructor
+@Getter
+@Setter
```
여기 @Setter가 들어가야하는 이유가 있을까요?

SSUHYUNKIM commented on src/main/java/com/ktb/paperplebe/common/service/CookieService.java:
```diff
@@ -0,0 +1,34 @@
+        ResponseCookie deleteCookie = ResponseCookie.from(cookieName, "")
+                .path("/")
+                .httpOnly(true)
+                .maxAge(0)
```
중복으로 2번 들어간 이유를 알 수 있을까요? 이전 프로젝트에서 오류가 있어서 중복으로 했었던 거 같은데 정확한 이유가 기억이 안나네요..

cjm0611 replied:
엇 실수로 중복으로 적었네요. postman으로 테스트 할 때는 쿠키가 정상적으로 delete되는 것으로 확인하여 우선 중복 제거했습니다.

SSUHYUNKIM commented on src/main/java/com/ktb/paperplebe/oauth/OAuthAttributes.java:
```diff
@@ -0,0 +1,43 @@
+    private final String nameAttributeKey;
+    private final OAuth2UserInfo oauth2UserInfo;
+
+    @Builder
```
혹시 Builder가 어디서 사용되고 있는 건가요?

cjm0611 replied:
빌더 패턴을 사용하고 있지 않아 제거했습니다!

SSUHYUNKIM commented on src/main/java/com/ktb/paperplebe/oauth/constant/SocialType.java:
```diff
@@ -0,0 +1,21 @@
+        return Arrays.stream(values())
+                .filter(type -> type.id.equals(registrationId))
+                .findFirst()
+                .orElseThrow(NoSuchElementException::new);
```
에러 메세지를 추가하는건 어떤가요?

SSUHYUNKIM commented:
저희 소셜 로그인 카카오만 제공하기로 했었나요?

SSUHYUNKIM commented on src/main/java/com/ktb/paperplebe/user/controller/UserController.java:
```diff
@@ -0,0 +1,35 @@
+    @Secured(UserRole.ROLE_USER_VALUE)
+    @GetMapping()
+    public UserInfoResponse getUserInfo(
+            @AuthenticationPrincipal Long id
```
정확히 어떤 id인지 명시하는건 어떤가요?

SSUHYUNKIM commented on src/main/java/com/ktb/paperplebe/user/service/UserService.java:
```diff
@@ -0,0 +1,44 @@
+
+    public User findById(Long id) {
+        return userRepository.findById(id)
+                .orElseThrow(() -> new RuntimeException("Invalid user ID: " + id));
```
나중에 exception 패키지를 만들어서 예외 처리를 작업하는 건 어떤가요?

cjm0611 replied:
저도 exception 패키지를 따로 만들까 고민했는데 바로 캐치해서 말씀해주시네요 👍
나중에 리팩터링할 때 작업할까요, 아니면 지금 진행할까요?

SSUHYUNKIM replied:
저도 예외 처리할 부분들이 있고 전역으로 처리해야할 부분도 존재할거 같아서 브랜치 새로 파는게 좋을 거 같아요!

cjm0611 commented:
> 저희 소셜 로그인 카카오만 제공하기로 했었나요?

구체적인 사항은 정하지 않았던 것 같습니다. 디자인에 카카오만 있어서 우선 카카오만 작업했는데, 다른 소셜 타입 추가하는 것은 공수가 크게 들지 않을 것 같아요! 대신 필수 입력 사항(프로필 이미지 등)을 정한다면 그 부분은 소셜 별로 따로 확인이 필요할 것 같습니다.

cjm0611 committed:
docs: api 문서 구체화

cjm0611 committed:
feat: CORS 허용 도메인 설정과 허용 및 노출 헤더 변경

- @Value를 사용하여 설정 파일에서 허용된 도메인을 가져오도록 변경
- 허용 및 노출 헤더를 업데이트

cjm0611 committed:
fix: jwt 토큰 검사 로직 수정

만료된 JWT 토큰일 경우 validation false 반환
access token이 유효하지 않은 경우 refresh token 검사 후 새로운 토큰 반환

cjm0611 committed:
feat: 사용하지 않는 setter annotation 삭제

cjm0611 committed:
refactor: 불필요한 빌드 패턴 제거

cjm0611 committed:
refactor: 명확한 네이밍인 userId로 변경

cjm0611 committed:
feat: 소셜 로그인 시 프로필 이미지 링크 가져오기

cjm0611 commented:
회원가입 시 프로필 이미지를 필수로 받도록 설정했습니다.

cjm0611 committed:
fix: 중복으로 쿠키를 만료시키던 로직 삭제

cjm0611 committed:
refactor: accessToken이 유효하지 않고 refreshToken이 유효한 경우 코드 간략화

cjm0611 committed:
fix: 토큰 갱신 로직 간략화

cjm0611 committed:
refactor: 토큰 만료, 삭제 로직 추가

토큰이 만료되었는지 확인하는 메서드 삭제

cjm0611 committed:
refactor: 잘못된 소셜 타입에 대한 에러 메세지 변경

SSUHYUNKIM reviewed (approved):
수고하셨습니다!

cjm0611 committed:
refactor: cors bean 이름 구분
